File: tomcat/src/spg/models/Library.java
package spg.models;

import org.json.JSONObject;

/**
 * Model for a Library.
 * 
 * @author Paul Barnhill, Kyle Tran, Carl Clermont
 * @version 2018-04-18
 */
public class Library {
	private final String libSiglum;
	private final String countryID;
	private final String city;
	private final String library;
	private final String address1;
	private final String address2;
	private final String postCode;
	
	public Library(String libSiglum, String countryID, String city, String library,
			String address1, String address2, String postCode) {
		this.libSiglum = libSiglum;
		this.countryID = countryID;
		this.city = city;
		this.library = library;
		this.address1 = address1;
		this.address2 = address2;
		this.postCode = postCode;
	}
	
	public String getlibSiglum() {
		return this.libSiglum;
	}
	
	public String getCountryID() {
		return this.countryID;
	}
	
	public String getCity() {
		return this.city;
	}
	
	public String getLibrary() {
		return this.library;
	}
	
	public String getAddress1() {
		return this.address1;
	}
	
	public String getAddress2() {
		return this.address2;
	}
	
	public String getPostCode() {
		return this.postCode;
	}
	
	/**
	 * Returns a JSONObject representation of this Country
	 */
	public JSONObject toJSON() {
		JSONObject j = new JSONObject();
		j.put("libSiglum", this.libSiglum);
		j.put("countryID", this.countryID);
		j.put("city", this.city);
		j.put("library", this.library);
		j.put("address1", this.address1);
		j.put("address2", this.address2);
		j.put("postCode", this.postCode);
		return j;
	}
	
	/**
	 * Returns a JSONObject String representation of this Country.
	 */
	public String toString() {
		return this.toJSON().toString();
	}
}

File: node/src/models/library.ts
/**
 * @author Paul Barnhill
 * @version 2018-04-16
 */
import { Country } from '@src/models/country';
import SpgModel from '@src/models/SpgModel.ts';
import SpgProxy from '@src/proxies/SpgProxy.ts';
import proxyFactory from '@src/proxies/ProxyFactory.ts';

/**
 * Properties of a Library.
 */
export interface Properties {
	libSiglum: string;
	countryID: string;
	city: string;
	library: string;
	address1: string;
	address2: string;
	postCode: string;
	// Index signature to allow iteration over properties in Typescript
	[x: string]: any;
}

/**
 * Data model for a Library.
 */
export class Library extends SpgModel implements Properties {
	public readonly libSiglum: string;
	public countryID: string;
	public city: string;
	public library: string;
	public address1: string;
	public address2: string;
	public postCode: string;

	public static MAX_LENGTHS = {
		libSiglum: 10,
		countryID: 20,
		city: 255,
		library: 255,
		address1: 255,
		address2: 255,
		postCode: 12,
	};

	// Index signature to allow iteration over properties in Typescript
	[x: string]: any;

	public constructor(props: Properties) {
		super();

		for (let k in props) {
			if (k === 'libSiglum' || k === 'countryID' || k === 'city' || k === 'library') {
				if (!props[k]) {
					throw Error(k + ' cannot be empty');
				}
			}
			this[k] = props[k];
		}
	}

	/**
	 * Returns a copy of the property map of this Library.
	 */
	public toProperties(): Properties {
		return {
			libSiglum: this.libSiglum,
			countryID: this.countryID,
			city: this.city,
			library: this.library,
			address1: this.address1 || '',
			address2: this.address2 || '',
			postCode: this.postCode || '',
		};
	}

	/**
	 * Creates an empty Library property map.
	 * @return
	 */
	public static createProperties(): Properties {
		return {
			libSiglum: '',
			countryID: '',
			city: '',
			library: '',
			address1: '',
			address2: '',
			postCode: ''
		};
	}
}
